instrumentation/active_job/test/instrumentation/active_job/patches/active_job_callbacks_test.rb

```diff
@@ -0,0 +1,191 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+require 'test_helper'
+
+require_relative '../../../../lib/opentelemetry/instrumentation/active_job'
+
+describe OpenTelemetry::Instrumentation::ActiveJob::Patches::ActiveJobCallbacks do
+  let(:exporter) { EXPORTER }
+  let(:spans) { exporter.finished_spans }
+  let(:send_span) { spans.find { |s| s.name == 'TestJob send' } }
+  let(:process_span) { spans.find { |s| s.name == 'TestJob process' } }
+
+  before { exporter.reset }
+
+  describe 'perform_later' do
+    it 'traces enqueuing and processing the job' do
+      TestJob.perform_later
+
+      _(send_span).wont_be_nil
+      _(process_span).wont_be_nil
+    end
+  end
+
+  describe 'perform_now' do
+    it 'only traces processing the job' do
+      TestJob.perform_now
+
+      _(send_span).must_be_nil
+      _(process_span).wont_be_nil
+    end
+  end
+
+  describe 'span kind' do
+    it 'sets correct span kinds for inline jobs' do
+      TestJob.perform_later
+
+      _(send_span.kind).must_equal(:client)
+      _(process_span.kind).must_equal(:server)
+    end
+
+    it 'sets correct span kinds for all other jobs' do
+      # Change the queue adapter so we get the right behavior
+      ::ActiveJob::Base.queue_adapter = :async
+
+      TestJob.perform_later
+      # We need to ensure that the no-op Test Job actually runs in the background thread
+      sleep 1
```

Comment:
> 💭 I will do some research and see if we can test this without sleeps. I am concerned that this may result in non-deterministic tests.

That's certainly possible. Given that we specifically change behavior (to follow semantic conventions) to see if we're in a queue that executes in a separate thread (ie: not a test or inline job adapter), I don't have an easy answer here. We *could* just skip testing those bits, if we're concerned about the flaky tests or don't care all that much.

Reply:
I fixed this up by shutting down the async executor thread in the tests where we switch to that adapter - there's no sleep required in that case. There's still _some_ chance for flakiness if the thread doesn't shut down correctly, but I think it's pretty drastically reduced.

I looked further into using the `:test` adapter, but it really expects to be running in a rails test suite. I started running into issues and realized I wasn't really saving any hassle at that point.